File: py/src/aoc2020/day6.py
import fileinput


def part1(lines):
    '''
    >>> part1(['abcx', 'abcy', 'abcz'])
    6
    >>> part1(['abc', '', 'a', 'b', 'c', '', 'ab', 'ac', '', 'a', 'a', 'a', 'a', '', 'b'])
    11
    '''
    total, group = 0, set()
    for line in lines:
        row = set(line.strip())
        if row:
            group.update(row)
        else:
            total += len(group)
            group = set()
    return total + len(group)


def part2(lines):
    '''
    >>> part2(['abc', '', 'a', 'b', 'c', '', 'ab', 'ac', '', 'a', 'a', 'a', 'a', '', 'b'])
    6
    '''
    total, group = 0, None
    for line in lines:
        row = set(line.strip())
        if row:
            if group is None:
                group = row
            else:
                group.intersection_update(row)
        elif group is not None:
            total += len(group)
            group = None
    return total + len(group or ())


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))

File: rs/src/day6.rs
pub fn part1<'a, I, S>(lines: I) -> u32
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut sum = 0u32;
    let mut group = 0u32;
    for line in lines.into_iter() {
        let row = line.as_ref().bytes().fold(0u32, |a, b| a | 1 << (b & 31));
        if row == 0 {
            sum += group.count_ones();
            group = 0;
        } else {
            group |= row;
        }
    }
    sum + group.count_ones()
}

pub fn part2<'a, I, S>(lines: I) -> u32
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut sum = 0u32;
    let mut group = u32::MAX;
    for line in lines.into_iter() {
        let row = line.as_ref().bytes().fold(0u32, |a, b| a | 1 << (b & 31));
        if row == 0 {
            sum += group.count_ones();
            group = u32::MAX;
        } else {
            group &= row;
        }
    }
    sum + group.count_ones()
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &[&str] = &[
        "abc", "", "a", "b", "c", "", "ab", "ac", "", "a", "a", "a", "a", "", "b",
    ];

    #[test]
    fn part1_examples() {
        assert_eq!(6, part1(&["abcx", "abcy", "abcz"]));
        assert_eq!(11, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(6, part2(EXAMPLE));
    }
}
